lib/db/lowlevel.go

```diff
@@ -590,6 +627,123 @@ func (db *Lowlevel) gcBlocks() error {
 	return db.Compact()
 }
 
+func (db *Lowlevel) gcDeleted() error {
+	// The deleted file GC finds items that are deleted and older than
+	// deletedGCAge (default 180 days). If we are in sync on this item (no
+	// need entry) we remove it from the database, along with the
+	// corresponding remote entries.
+	//
+	// If the remote side doesn't GC deleted files then the entry will at
+	// some point pop up again. Ideally this should therefore be
+	// symmetrically configured.
+
+	if deletedGCAge <= 0 {
+		// <= 0 is explicitly disabled
+		return nil
+	}
+	if deletedGCAge < deletedGCMinAge {
+		// an explicit but not reasonable value, we adjust upwards a bit.
+		deletedGCAge = deletedGCMinAge
+	}
+
+	db.gcMut.Lock()
+	defer db.gcMut.Unlock()
+
+	t, err := db.newReadWriteTransaction()
+	if err != nil {
+		return err
+	}
+	defer t.Release()
+
+	cutoff := time.Now().Add(-deletedGCAge)
+
+	it, err := db.NewPrefixIterator([]byte{KeyTypeSequence})
+	if err != nil {
+		return err
+	}
+	defer it.Release()
+	for it.Next() {
+		// Load our version of the file. See if it's an old deleted one.
+
+		dkey := deviceFileKey(it.Value())
+		f, ok, err := t.getFileTrunc(dkey, true)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			continue
+		}
+		if !f.IsDeleted() || f.ModTime().After(cutoff) {
+			continue
+		}
+
+		name := []byte(f.FileName())
+		folder, ok := t.keyer.FolderFromDeviceFileKey(dkey)
+		if !ok {
+			continue
+		}
+
+		// See if we have a need entry for this file. In that case we
+		// shouldn't just drop it.
+
+		needKey, err := t.keyer.GenerateNeedFileKey(nil, folder, name)
+		if err != nil {
+			return err
+		}
+		_, err = t.Get(needKey)
+		if err == nil {
+			// need it
+			continue
+		}
+		if !backend.IsNotFound(err) {
+			return err
+		}
+
+		// Load the global list for this file.
+
+		gvKey, err := t.keyer.GenerateGlobalVersionKey(nil, folder, name)
+		if err != nil {
+			return err
+		}
+		gvBs, err := t.Get(gvKey)
+		if err != nil {
+			return err
+		}
+		var vl VersionList
+		if err := vl.Unmarshal(gvBs); err != nil {
+			return err
+		}
+
+		// We can clean this entry.
```

Comment:
Shouldn't we check that noone else needs this file either (i.e. it's deleted for everyone)? I guess it's mostly theoretical, but in principle a device could be offline since before the file was deleted for such a long time that GC kicks in, and whenever it comes online, it will propagate it's file instead of delete it.

Reply:
Yes

Reply:
Yes, I even did this to begin with. It seems reasonable to do so. In practice, in *complicated* setups, it seems there is always this device that's still in the config even though it was decommisioned six months ago and who has no modern index entries... You could argue that this means this feature does not work and that's OK and I have no excellent counter argument to that.

Reply:
Well surely they can just remove it?
We also have last seen for the device which we could use.

Reply:
Yep. Solutions abound. This is the smaller problem I think, the other one is more serious.